ProjFS.Mac/PrjFSLib/PrjFSLib.cpp

```diff
@@ -94,10 +97,27 @@ static PrjFS_Callbacks s_callbacks;
 static dispatch_queue_t s_messageQueueDispatchQueue;
 static dispatch_queue_t s_kernelRequestHandlingConcurrentQueue;
 
-// Map of relative path -> set of pending message IDs for that path, plus mutex to protect it.
-static unordered_map<string, set<uint64_t>> s_PendingRequestMessageIDs;
-static mutex s_PendingRequestMessageMutex;
+struct CaseInsensitiveStringCompare
+{
+    bool operator() (const std::string& lhs, const std::string& rhs) const
+    {
+        return strcasecmp(lhs.c_str(), rhs.c_str()) < 0;
+    }
+};
+
+struct MutexAndUseCount
+{
+    shared_ptr<mutex> mutex;
+    int useCount;
+};
 
+// Map of relative path -> MutexAndUseCount for that path, plus mutex to protect the map itself.
+typedef map<string, MutexAndUseCount, CaseInsensitiveStringCompare> PathToMutexMap;
```

Comment:
@pmj do you know if we need to worry about being case insensitive?  Or will the kext always use the same\actual case when sending paths to user-mode?

Reply:
CC @sanoursa  as well

Reply:
I'm not sure; the path is ultimately coming from `vn_getpath()` and I believe it gets its information from the name cache. The question is therefore if the name cache is populated with canonical names or the component names that came in from user space.

One correct way to check whether 2 paths refer to the same file would be to `stat()` them both and check if fsid & inodes match. The kext should already have this information, so perhaps we should be sending that tuple along with each kernel -> user message and using that as our map key?

Given that macOS does support case-sensitive file systems, case-insensitive string comparison is not really ideal here. I mean, we can of course declare that we won't support VFS for Git on case-sensitive file systems, but I'd still be worried about edge cases.

Reply:
Incidentally, the inode approach also tackles any hardlink issues.

Reply:
> One correct way to check whether 2 paths refer to the same file would be to stat() them both and check if fsid & inodes match. The kext should already have this information, so perhaps we should be sending that tuple along with each kernel -> user message and using that as our map key?

That's a good idea, I'll see about plumbing that through.

Reply:
> check if fsid & inodes match

@pmj Totally unrelated to this PR, your comment reminded me of something I'd been thinking about related to how the kext could handle a failures of `vn_getpath` (or perhaps it wouldn't even call it at all).

Do you have any thoughts an approach where the kext passes fsid & inode and then the user-mode library calls `fsgetpath`?  I wasn't sure if that would potentially be more reliable than `vn_getpath`.